### PR TITLE
feat(rome_js_analyze): `noParameterProperties`

### PR DESCRIPTION
**Pull request by Conaclos**
## Summary

This implements *ESLint*'s [no-parameter-properties](https://typescript-eslint.io/rules/no-parameter-properties/).

## Test Plan

Unit tests and doc-tests included.

### CODE REVIEW

**Comment by ematipico** [comment]
The documentation says that this rule is deprecated

**Comment by Conaclos** [comment]
> The documentation says that this rule is deprecated

Yes, the rule was replaced by [parameter-properties](https://typescript-eslint.io/rules/parameter-properties).
However, this does not fit with the philosophy of Rome (`no` or `use`, and do not provide configuration when it is not necessary).

**Comment by Conaclos** [comment]
By the way I tried to impelment an action to turn the parameter property in to a regular property.
However, I encountered an issue: when I mutate a node, I cannot mutate a descendant without invalidating the previous mutation.
It is particularly inconvenient.

I have to figure out how to overpass this limitation.

**Comment by ematipico** [comment]
I have some reservations about this rule, here's my concerns:

- if and once this rule is promoted, which group would this rule fit?
- there are frameworks that use these properties, such as NestJS: https://docs.nestjs.com/providers#services, wouldn't this rule create friction? 
- are there any other motivations other than the the confusing part? I used parameter properties inside constructors many times, wouldn't this rule create friction?

**Comment by Conaclos** [comment]
> if and once this rule is promoted, which group would this rule fit?

I think that `Style` could be a good fit.

> there are frameworks that use these properties, such as NestJS

I was not aware of this use. It could definitively bring friction.

> are there any other motivations other than the the confusing part?

The arguments I could add:
- TypeScript should avoid code generation. Parameter properties are syntactic sugars. Yes, TypeScript has other features that generate code (enums, namespaces). However, namespaces are no longer recommended, and enums are popular and could even be one day part of JavaScript.
- It is not possible to declare a private parameter property (using she-bang #).

**Comment by leops** [comment]
> By the way, I tried to implement an action to turn the parameter property into a regular property.
> 
> However, I encountered an issue: when I mutate a node, I cannot mutate a descendant without invalidating the previous mutation. It is particularly inconvenient.
> 
> I have to figure out how to overpass this limitation.

I think the easiest way to implement this without encountering conflicts in the mutation would be to:
- Remove the `TsPropertyParameter` from its parent `JsConstructorParameterList` using `splice_slots`
- Replace the old constructor parameter list with the modified list in the `JsClassMemberList` containing the constructor using `replace_child`
- Insert the new `JsPropertyClassMember` node in the modified class member list using `splice_slots`
- Queue an operation to replace the old class member list with the modified list in the `BatchMutation` by calling `replace_node`

**Comment by xunilrj** [comment]
@Conaclos  Do you have a version with the mutation problem so I can look it up?
I think we may have a version of `BatchMutation` that fix this, but it was never merged because it is slower.

**Comment by Conaclos** [comment]
@xunilrj 
I think that the action could be made in a separate PR.
I could like to propose the PR in its current state.

File: crates/rome_js_analyze/src/analyzers/nursery/no_parameter_properties.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+use rome_js_syntax::TsPropertyParameter;
+use rome_rowan::AstNode;
+
+declare_rule! {
+    /// Disallow the use of parameter properties in class constructors.
+    ///
+    /// TypeScript includes a "parameter properties" shorthand for declaring a class constructor parameter and class property in one location.
+    /// Parameter properties can be confusing to those new to TypeScript as they are less explicit than other ways of declaring and initializing class members.
+    /// Moreover, private class properties (properties starting with `#`) cannot be turn into "parameter properties".
+    /// This questions the future of this feature.
```
Comment (ematipico) [inline crates/rome_js_analyze/src/analyzers/nursery/no_parameter_properties.rs:13]:
```suggestion
    /// Disallow the use of parameter properties in class constructors.
    ///
    /// TypeScript includes a "parameter properties" shorthand for declaring a class constructor parameter and class property in one location.
    /// Parameter properties can confuse those new to TypeScript as they are less explicit than other ways of declaring and initializing class members.
    /// Moreover, private class properties, starting with `#`, cannot be turned into "parameter properties".
    /// This questions the future of this feature.
```

File: crates/rome_js_analyze/src/analyzers/nursery/no_parameter_properties.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        Some(())
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, _: &Self::State) -> Option<RuleDiagnostic> {
+        let param_prop = ctx.query();
+        Some(RuleDiagnostic::new(
+            rule_category!(),
+            param_prop.range(),
+            markup! {
+                "Unexpected "<Emphasis>"parameter property"</Emphasis>"."
```
Comment (ematipico) [inline crates/rome_js_analyze/src/analyzers/nursery/no_parameter_properties.rs:58]: We should rephrase the message here. Just saying "unexpected parameter property" seems like poor messaging from our end. While it's true that we add a note later, **editors** don't have the privilege of seeing that note.

Few suggestions: "This parameter property can confuse.", "This parameter is not explicit enough.", and "Avoid parameter properties and rely on its code generation."

**Review by ematipico — APPROVED** [review]

**Pull request merged**